lib/read.js

```diff
@@ -139,11 +140,131 @@ class ReadSql extends AdapterRead {
+    }
+
+    //Query execution
+
+    read(from, to, limit, state) {
+        var queryExecutionPromise;
+
+        this.queueSize = limit;
+
+        if (this.raw) {
+            queryExecutionPromise = this.getRawResult();
+        } else {
+            var query = this.getPaginationQuery(from, to);
+
+            if (this.debugOption) {
+                queryExecutionPromise = this.handleDebug(query);
+            } else if (this.optimization_info.reduce_every) {
+                queryExecutionPromise = this.getReduceEveryResults(query, from, to);
+            } else {
+                queryExecutionPromise = this.executeQuery(query, to);
+            }
+        }
+
+        return queryExecutionPromise
+        .catch((err) => {
+            if (/Pool (is|was) destroyed/.test(err.message)) {
+                var connectionInfo = this.baseQuery.client.connectionSettings;
+                throw errors.runtimeError('RT-INTERNAL-ERROR', {
+                    error: 'could not connect to database: ' + JSON.stringify(connectionInfo)
+                });
+            } else {
+                throw err;
+            }
+        });
+    }
+
+    // raw query
+
+    getRawResult() {
+        return this.baseQuery
+        .then((res) => {
+            return this.handleRawResponse(res);
+        })
+        .then((points) => {
+            return {
+                points: this.formatPoints(points),
+                readEnd: new JuttleMoment(Infinity)
+            };
+        });
+    }
+
+    //the raw query response format can be different based on client
+    handleRawResponse(points) {
+        return points;
+    }
+
+    getPaginationQuery(from, to) {
+        var query = this.baseQuery.clone();
+
+        query = query.limit(this.getNextLimit());
+
+        if (this.timeField) {
+            if (from) {
+                query = query.where(this.timeField, '>=', new Date(from.valueOf()));
+                this.logger.debug('start time', new Date(from.valueOf()).toISOString());
+            }
+            if (to) {
+                query = query.where(this.timeField, '<', new Date(to.valueOf()));
+                this.logger.debug('end time', new Date(to.valueOf()).toISOString());
+            }
+        } else {
+            query = query.offset(this.offsetCount);
+        }
+        return query;
+    }
+
+    // debug option handler
+
+    handleDebug(query) {
+        this.logger.debug('returning query string as point');
+        return Promise.try(() => {
+            return  {
+                points: [{ query: query.toString() }],
+                readEnd: new JuttleMoment(Infinity)
+            };
+        });
+    }
+
+    // "reduce -every" batch query
+
+    getReduceEveryResults(original_query, from, to) {
+        var timeField = this.timeField || 'time';
+        var buckets = this.getBatchBuckets(from, to);
+
+        return Promise.mapSeries(buckets.slice(1), (bucket, i, len) => {
```

Comment:
Also why the `slice(1)`? Should have `getBatchBuckets` return what you really want instead of dropping the first bucket here.